torch/csrc/utils/tensor_numpy.cpp

```diff
@@ -227,9 +230,12 @@ ScalarType numpy_dtype_to_aten(int dtype) {
       ((PyTypeObject*)pytype.get())->tp_name);
 }
 
+bool is_numpy_int(PyObject* obj) {
+  return PyArray_IsIntegerScalar(obj) && ! PyBool_Check(obj);
```

Comment:
So the question I have right now is: what is the root reason we have to guard against bool here? I understand that the tests pass with it and they don't without it, but that doesn't say why.

Reply:
This `is_numpy_int` function was added to add numpy support for THPUtils_checkLong, and that's really the only reason it's here. The implementation for checkLong on master was `return PyLong_Check(obj) && !PyBool_Check(obj);`. 

If we allow checkLong to treat booleans as long, other code breaks. It's hard to track down everything it would break, but presumably it's because of one of these if/else statements checking types of parameters.

Reply:
Does PyArray_IsIntegerScalar return True for Py_Bool?

Reply:
It looks like the answer to that is Yes, because it calls PyLong_Check, which returns True for bool. 

if you’re checking that something is an np int then the best api to use looks like
 https://github.com/numpy/numpy/blob/1783e6fc437f34af1d652cbae36a68abcbf41fbc/numpy/core/include/numpy/ndarrayobject.h#L53 , not PyArray_IsIntegerScalar.

Reply:
> Does PyArray_IsIntegerScalar return True for Py_Bool?

Yes because `bool.__flags__ & (1 << 24) != 0` as we looked at before.

FYI, on master, related to the np.bool vs np.bool_ that you pointed out, I'm not trying to change this particular behavior in this PR:

```
>>> torch.pow(torch.ones(4), True)
tensor([1., 1., 1., 1.])
>>> torch.pow(torch.ones(4), np.array([True])[0])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: pow() received an invalid combination of arguments - got (Tensor, numpy.bool_), but expected one of:
```

Reply:
> FYI, on master, related to the np.bool vs np.bool_ that you pointed out, I'm not trying to change this particular behavior in this PR:

I agree that we're not trying to change this behavior. In the np.bool vs np.bool_ thread we had going, I asked what the behavior of this function was on np.bool_.

I think that instead of writing `return PyArray_IsIntegerScalar(obj) && ! PyBool_Check(obj);` we should write return `PyArray_IsScalar((obj), Integer))`, like in here: https://github.com/numpy/numpy/blob/1783e6fc437f34af1d652cbae36a68abcbf41fbc/numpy/core/include/numpy/ndarrayobject.h#L52-L53

Otherwise, the semantics of `is_numpy_int` become too confusing: `is_numpy_int(PyBool)` is False right now, but `is_numpy_int(PyLong)` is True. I think what we really want is:
- is_numpy_int(Python number type) -> False
- is_numpy_int(numpy integer type) -> True

Reply:
ah, yes that makes sense, I'll try that.